relay-general/src/pii/attachments.rs

```diff
@@ -71,43 +65,283 @@ fn apply_regex_to_bytes(
         return false;
     }
 
-    const DEFAULT_PADDING: u8 = b'x';
-    const MASK_PADDING: u8 = b'*';
+    for (start, end) in matches {
+        data[start..end].apply_redaction(&rule.redaction);
+    }
+    true
+}
 
-    match rule.redaction {
-        Redaction::Default | Redaction::Remove => {
-            for (start, end) in matches {
-                for c in &mut data[start..end] {
-                    *c = DEFAULT_PADDING;
+fn apply_regex_to_utf16le_bytes(
+    data: &mut [u8],
+    rule: &RuleRef,
+    regex: &Regex,
+    replace_behavior: &ReplaceBehavior,
+) -> bool {
+    let mut changed = false;
+    for segment in MutSegmentIter::new(data, *UTF_16LE) {
+        let segment_wstr = unsafe { WStr::from_utf16le_unchecked_mut(segment.raw) };
+
+        match replace_behavior {
+            ReplaceBehavior::Value => {
+                for re_match in regex.find_iter(&segment.decoded) {
+                    changed = true;
+                    let match_wstr = get_wstr_match(&segment.decoded, re_match, segment_wstr);
+                    match_wstr.apply_redaction(&rule.redaction);
                 }
             }
-        }
-        Redaction::Mask => {
-            for (start, end) in matches {
-                let match_slice = &mut data[start..end];
-                for c in match_slice.iter_mut() {
-                    *c = MASK_PADDING;
+            ReplaceBehavior::Groups(ref replace_groups) => {
+                for captures in regex.captures_iter(&segment.decoded) {
+                    for group_idx in replace_groups.iter() {
+                        if let Some(re_match) = captures.get(*group_idx as usize) {
+                            changed = true;
+                            let match_wstr =
+                                get_wstr_match(&segment.decoded, re_match, segment_wstr);
+                            match_wstr.apply_redaction(&rule.redaction);
+                        }
+                    }
                 }
             }
         }
-        Redaction::Hash => {
-            for (start, end) in matches {
-                let hashed = hash_value(&data[start..end]);
-                replace_bytes_padded(hashed.as_bytes(), &mut data[start..end], DEFAULT_PADDING);
+    }
+    changed
+}
+
+/// Extract the matching encoded slice from the encoded string.
+fn get_wstr_match<'a>(all_text: &str, re_match: Match, all_encoded: &'a mut WStr) -> &'a mut WStr {
+    let mut encoded_start = 0;
+    let mut encoded_end = 0;
```

Comment:
I think this needs to be initialized with `all_encoded.len()`. Otherwise, if the end index matches the end of the string, this will remain `0` as the loop below never visits this location.